Add support for rewrite-target annotation

ausmith opened the pull request:
#55 

Adds support for ingress annotation: `ingress.kubernetes.io/rewrite-target`.

Our test case for this work was `ingress.kubernetes.io/rewrite-target: /` as that has been the most common use case for us internally. This implementation SEEMS to be feature similar to the current implementation in the [kubernetes/ingress](https://github.com/kubernetes/ingress) nginx ingress controller, but our test cases were rather limited so unsure.

Used https://gist.github.com/kushmansingh/47644ef4db96f11b1ea450ae19072552 to do our testing but could have used most any other container that echoes the headers.

ausmith commented:
Have now also tested `ingress.kubernetes.io/rewrite-target: /yourserver` and doing a curl against the kube cluster with this gives the following results:

```
curl http://test-ingress.com/myserver
# output: 2017/10/03 15:57:54  /yourserver

curl http://test-ingress.com/myserver/xyz
# output: 2017/10/03 15:57:54  /yourserver
```

With `ingress.kubernetes.io/rewrite-target: /yourserver` we get:

```
curl http://test-ingress.com/myserver
# output: 2017/10/03 16:00:51  /yourserver/

curl http://test-ingress.com/myserver/xyz
# output: 2017/10/03 16:00:25  /yourserver/xyz
```

So it is the responsibility of the ingress definer to include the trailing slash in the annotation or they might get unexpected responses.

jcmoraisjr commented on rootfs/etc/haproxy/template/haproxy.tmpl:
```diff
@@ -284,14 +296,25 @@ frontend httpsfront-{{ $host }}
 {{ range $location := $server.Locations }}
 {{ if ne $location.Proxy.BodySize "" }}
-    use_backend error413 if { path_beg {{ $location.Path }} } { req.body_size gt {{ sizeSuffix $location.Proxy.BodySize }} }
+    use_backend error413 if {{ $location.Backend }}_path_match { req.body_size gt {{ sizeSuffix $location.Proxy.BodySize }} }
```
acl should be declared before being used.

jcmoraisjr commented on rootfs/etc/haproxy/template/haproxy.tmpl:
```diff
@@ -108,6 +108,18 @@ frontend httpfront
+{{ if ne $server.Alias "" }}
+    reqrep ^([^\ :]*)\ {{ $location.Path }}[/]?(.*) \1\ {{ $rewriteTarget }}\2 if {{ $location.Backend }}_path_match ishost-{{ $server.HostnameLabel }} or ishost-{{ $server.HostnameLabel }}-alias
+{{ else }}
+    reqrep ^([^\ :]*)\ {{ $location.Path }}[/]?(.*) \1\ {{ $rewriteTarget }}\2 if {{ $location.Backend }}_path_match ishost-{{ $server.HostnameLabel }}
```
These lines are almost the same, use an inline `if` instead.

Prefer http-request if possible, otherwise declare them after all http-requests in order to avoid warning from haproxy cfg parser.

ausmith replied:
Unfortunately, the documentation indicates `http-request` does not support capture groups, hence we had to fall back to the `reqrep` option instead.

jcmoraisjr commented on rootfs/etc/haproxy/template/haproxy.tmpl:
```diff
@@ -108,6 +108,18 @@ frontend httpfront
 {{ if $server.UseHTTP }}
 {{ range $location := $server.Locations }}
+    http-request set-var(req.ingress_orig_path) path
+    acl {{ $location.Backend }}_path_match var(req.ingress_orig_path) -m beg {{ $location.Path }} if or ishost-{{ $server.HostnameLabel }} ishost-{{ $server.HostnameLabel }}-alias
```
Add `ishost-...-alias` inside a condition - ne Alias "".

Prefer loop acls+setvar in a separate loop of reqrep. Worst from the point of view of the template but better to the final cfg.

jcmoraisjr commented on rootfs/etc/haproxy/template/haproxy.tmpl:
```diff
@@ -108,6 +108,18 @@ frontend httpfront
 {{ range $server := $ing.HAServers }}
 {{ if $server.UseHTTP }}
 {{ range $location := $server.Locations }}
+    http-request set-var(req.ingress_orig_path) path
```
my assumption is that you are using a var because the path is being changed before match on use_backend, is it right? Note that you are using the same var name to different hosts and paths.

ausmith replied:
Right, I was finding that after the `reqrep` we could not execute the same `acl` against the path when choosing the backend. OH! Probably do need to move the variable setting to outside the loop though, good call there.

ausmith commented:
I believe I've handled each of the comments and have done local testing to verify that these changes still behave properly. 👍

jcmoraisjr commented:
Some other changes:

* Why use `[/]?(.*)` instead of just `(.*)`?
* Why use an acl instead change `path_beg` to `var(req.path) -m beg`? Note that if you have more than one location in the same server you'll have name collision
* A shorter var name like `req.path` would be fine
* `app-root` annotation also match `path` and should be changed
* There is also an ugly workaround named `HAMatchPath` (`config.go`) which also match `path_beg` and should also be changed
* `http-request` is still after a `rsprep` on https proxy - move the `http-request` in order to avoid a haproxy warn
* Try to rebase your next changes and the old commits into a single new commit - or leave it as it is if thing goes wrong

ausmith commented:
@jcmoraisjr ty for the feedback! I did a rebase with all of it in one commit as requested. Notes from your other bullets:
- Why use `[/]?(.*)` instead of just `(.*)`?
  - I have tested without the `[/]?(.*)` and it causes a redirect instead of the expected behavior if the path is not an exact match (i.e. `/myserver/` instead of `/myserver`). The trailing slash allows all of `/myserver`, `/myserver/`, and `/myserver/xyz` to match.
- Why use an acl instead change `path_beg` to `var(req.path) -m beg`? Note that if you have more than one location in the same server you'll have name collision
  - I have switched them all back to non-ACLs. I had read that using anonymous conditions meant they were compiled at runtime rather than at config load time somewhere, so unsure about the performance difference between ACLs and anonymous conditions.
- A shorter var name like `req.path` would be fine
  - I did this as you are totally right.
- `app-root` annotation also match `path` and should be changed
  - Fixed.
- There is also an ugly workaround named `HAMatchPath` (`config.go`) which also match `path_beg` and should also be changed
  - Fixed.
- `http-request` is still after a `rsprep` on https proxy - move the `http-request` in order to avoid a haproxy warn
  - Fixed.

jcmoraisjr commented:
Great, we're almost there:

Setvar is inside a loop (http proxy) but you need just one.

I'm worried about the regex on behalf of consistency. I couldn't reproduce the redirect, perhaps your backend service did it? On the other hand `[/]?` is removing a slash from the url and I'm ending up with (where path is `/abc` and rewrite is`/xpto`):

Ok:

* `/abc` -> `/xpto`
* `/abcd` -> `/xptod`

Not ok:

* `/abc/` -> `/xpto`
* `/abc/d` -> `/xptod`

> I have tested without the `[/]?(.*)` and ...

Did you mean `[/]?`? This was what I mean.

ausmith commented:
So, I have figured out the rewrite target issue overwriting `/` unexpectedly and added a caveat to the README about rewrite target not supporting a trailing `/`. Does that cover all cases now?

ausmith commented:
Only change since last comment is I found and fixed the source of this warning:
```
[WARNING] 279/130942 (98) : parsing [/etc/haproxy/haproxy.cfg:194] : a 'http-request' rule placed after a 'reqxxx' rule will still be processed before.
[WARNING] 279/130942 (98) : parsing [/etc/haproxy/haproxy.cfg:195] : a 'http-request' rule placed after a 'reqxxx' rule will still be processed before.
```
Required moving `http-request set-header X-Forwarded-Proto https` and the `set-var` that this PR adds above the `reqidel` on line 284 of the template.

jcmoraisjr commented:
Two small changes in the readme:

* Add `[1] only in snapshot tag` - just below `[0] canary` and change the tag of this new annotation from `[0]` to `[1]`
* Link the annotation to the description - have a look at how affinity was linked

And let me know what you think about this approach (not tested, just the idea):

`... {{ $location.Path }}(.*$)  \1\ {{ $rewriteTarget }}{{ if and (ne $rewriteTarget "/") (endsWith $location.Path "/")}}/{{ end }}\2  if...`

in order to:

* use only one `reqrep` (per frontend)
* change from _"unsupported trailing slash"_ to _"use the same pattern used in ingress path"_ - both ingress path and rewrite should, or should not, have the trailing slash

`endsWith` would be a new func to be implemented (template.go)

ausmith commented:
I have made the README adjustments accordingly.

The `reqrep` change might be possible, but I am running out of engineering time at work to make that tweak and thoroughly test it. What do you think about supporting a trailing slash on the `rewrite-target` annotation being a future enhancement @jcmoraisjr ? I will be happy to help continue expanding the `rewrite-target` capabilities down the road but do not have the `golang` experience to quickly knock out your suggestion.

jcmoraisjr commented:
Sounds good. The current implementation is good enough to be merged and we can evolve the config along with the 0.5 development. Btw thanks for helping us making this controller even better.
